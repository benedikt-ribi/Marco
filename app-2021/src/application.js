import "./master-detail.js";

/**
 * @return ClassMetadata
 * @param rootUrl {string}
 */
export function createAddressMetadata(rootUrl) {
  return {
    caption: "Address",
    pluralCaption: "Addresses",
    rootUrl: `${rootUrl}addresses`,
    getTitle: a => `${a.firstName} ${a.lastName}`,
    createNew: () => ({ firstName: "", lastName: "" }),
    properties: [
      {
        name: "firstName",
        type: "text",
        caption: "First Name",
      },
      {
        name: "lastName",
        type: "text",
        caption: "Last Name",
      },
    ],
  };
}

/**
 * @return ClassMetadata
 * @param rootUrl {string}
 */
export function createCarMetadata(rootUrl) {
  return {
    caption: "Car",
    pluralCaption: "Cars",
    rootUrl: `${rootUrl}cars`,
    getTitle: c => `${c.make} ${c.model}`,
    createNew: () => ({ make: "", model: "" }),
    properties: [
      {
        name: "make",
        type: "text",
        caption: "Make",
      },
      {
        name: "model",
        type: "text",
        caption: "Model",
      },
    ],
  };
}
