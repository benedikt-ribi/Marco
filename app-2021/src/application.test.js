/**
 * @jest-environment jsdom
 */

import { describe, expect, test } from "@jest/globals";
import "isomorphic-fetch";
import { testingRootUrl } from "./test-library";
import { application } from "./application";

describe("Application", () => {
  test("initialize", async () => {
    document.body.innerHTML = `
      <div id="listItems"></div>
      <div id="detail"></div>
    `;

    const listItems = document.getElementById("listItems");
    const detail = document.getElementById("detail");

    const app = new application();

    await app.initialize(listItems);

    expect(listItems.innerHTML).toMatchSnapshot();
    expect(detail.innerHTML).toMatchSnapshot();
  });

  test("clicking an item loads its details", async () => {
    document.body.innerHTML = `
      <div id="listItems"></div>
      <div id="detail"></div>
    `;

    const app = new application(testingRootUrl);
    const listItems = document.getElementById("listItems");
    const detail = document.getElementById("detail");

    await app.initialize(listItems);

    let links = listItems.getElementsByTagName("a");

    expect(links.length).toBe(12);

    const [firstLink] = links;

    expect(firstLink.textContent).toBe("Peter Smith");

    firstLink.click();

    expect(detail.innerHTML).toMatchSnapshot();
  });
});
