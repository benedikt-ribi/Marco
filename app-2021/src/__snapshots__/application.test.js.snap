// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application clicking an item loads its details 1`] = `"Peter Smith"`;

exports[`Application initialize 1`] = `"Only absolute URLs are supported"`;

exports[`Application initialize 2`] = `""`;
